app/src/animation/animation_control.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_animation_control
+
+#include <stdio.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/settings/settings.h>
+
+#include <drivers/animation.h>
+
+#include <zmk/animation.h>
+#include <zmk/animation/animation_control.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define PHANDLE_TO_DEVICE(node_id, prop, idx) DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, idx)),
+
+struct animation_control_work_context {
+    const struct device *animation;
+    struct k_work_delayable save_work;
+};
+
+struct animation_control_config {
+    const struct device **animations;
+    const size_t animations_size;
+    const uint8_t brightness_steps;
+    struct animation_control_work_context *work;
+    struct settings_handler *settings_handler;
+};
+
+struct animation_control_data {
+    bool active;
+    uint8_t brightness;
+    size_t current_animation;
+};
+
+static int animation_control_load_settings(const struct device *dev, const char *name, size_t len,
+                                           settings_read_cb read_cb, void *cb_arg) {
+#if IS_ENABLED(CONFIG_SETTINGS)
+    const char *next;
+    int rc;
+
+    if (settings_name_steq(name, "state", &next) && !next) {
+        if (len != sizeof(struct animation_control_data)) {
+            return -EINVAL;
+        }
+
+        rc = read_cb(cb_arg, dev->data, sizeof(struct animation_control_data));
+        if (rc >= 0) {
+            return 0;
+        }
+
+        return rc;
+    }
+
+    return -ENOENT;
+#else
+    return 0;
+#endif /* IS_ENABLED(CONFIG_SETTINGS) */
+}
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+static void animation_control_save_work(struct k_work *work) {
+    struct animation_control_work_context *ctx =
+        CONTAINER_OF(work, struct animation_control_work_context, save_work);
+    const struct device *dev = ctx->animation;
+
+    char path[40];
+    snprintf(path, 40, "%s/state", dev->name);
+
+    settings_save_one(path, dev->data, sizeof(struct animation_control_data));
+};
+
+static int animation_control_save_settings(const struct device *dev) {
+    const struct animation_control_config *config = dev->config;
+    struct animation_control_work_context *ctx = config->work;
+
+    k_work_cancel_delayable(&ctx->save_work);
+
+    return k_work_reschedule(&ctx->save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
+}
+#endif /* IS_ENABLED(CONFIG_SETTINGS) */
+
+int animation_control_handle_command(const struct device *dev, uint8_t command, uint8_t param) {
+    const struct animation_control_config *config = dev->config;
+    struct animation_control_data *data = dev->data;
+
+    switch (command) {
+    case ANIMATION_CMD_TOGGLE:
```

Comment:
It seems like the `param` value (which is set to the zone) is ignored for some of these commands. I assume this is intentional, but then why allow a zone to be provided to those macros within the behavior itself?

Reply:
As stated in the other comment, the `zone` param refers to the `animation_control` instance itself. Maybe we can think of some better naming or something 🤔.